contracts/hooks/fees/src/lib.rs

```diff
@@ -0,0 +1,233 @@
+pub enum ContractError {
+    #[error("{0}")]
+    Std(#[from] StdError),
+
+    #[error("{0}")]
+    PaymentError(#[from] cw_utils::PaymentError),
+
+    #[error("unauthorized")]
+    Unauthorized {},
+
+    #[error("hook paused")]
+    Paused {},
+}
+
+// version info for migration info
+pub const CONTRACT_NAME: &str = env!("CARGO_PKG_NAME");
+pub const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+pub const COIN_FEE_KEY: &str = "coin_fee";
+pub const COIN_FEE: Item<Coin> = Item::new(COIN_FEE_KEY);
+
+fn new_event(name: &str) -> Event {
+    Event::new(format!("hpl_hook_fee::{}", name))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    cw2::set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let owner = deps.api.addr_validate(&msg.owner)?;
+
+    hpl_ownable::initialize(deps.storage, &owner)?;
+    COIN_FEE.save(deps.storage, &msg.fee)?;
+
+    Ok(Response::new().add_event(
+        new_event("initialize")
+            .add_attribute("sender", info.sender)
+            .add_attribute("owner", owner)
+            .add_attribute("fee_denom", msg.fee.denom)
+            .add_attribute("fee_amount", msg.fee.amount),
+    ))
+}
+
+fn get_fee(deps: Deps) -> Result<FeeResponse, ContractError> {
+    let fee = COIN_FEE.load(deps.storage)?;
+
+    Ok(FeeResponse { fee })
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Ownable(msg) => Ok(hpl_ownable::handle(deps, env, info, msg)?),
+        ExecuteMsg::FeeHook(msg) => match msg {
+            FeeHookMsg::SetFee { fee } => {
+                COIN_FEE.save(deps.storage, &fee)?;
+
+                Ok(Response::new().add_event(
+                    new_event("set_fee")
+                        .add_attribute("fee_denom", fee.denom)
+                        .add_attribute("fee_amount", fee.amount),
+                ))
+            }
+            FeeHookMsg::Claim { recipient } => {
+                let owner = hpl_ownable::get_owner(deps.storage)?;
+                ensure_eq!(owner, info.sender, StdError::generic_err("unauthorized"));
+
+                let recipient = recipient.unwrap_or(owner);
+                let balances = deps.querier.query_all_balances(&env.contract.address)?;
+
+                let claim_msg: CosmosMsg = BankMsg::Send {
+                    to_address: recipient.into_string(),
+                    amount: balances,
+                }
+                .into();
+
+                Ok(Response::new()
+                    .add_message(claim_msg)
+                    .add_event(new_event("claim")))
+            }
+        },
+        ExecuteMsg::PostDispatch(_) => {
+            let fee = COIN_FEE.load(deps.storage)?;
```

Comment:
Should `COIN_FEE` contains only one denom?

Reply:
good question, I think for now, for simplicity, yes